sdk/include/platform/sunburst/platform-gpio.hh

```diff
@@ -0,0 +1,444 @@
+ * https://lowrisc.org/sonata-system/doc/ip/gpio.html
+ */
+struct SonataGpioBase : private utils::NoCopyNoMove
+{
+	uint32_t output;
+	uint32_t input;
+	uint32_t debouncedInput;
+	uint32_t outputEnable;
+
+	/**
+	 * The mask of bits of the `output` register that contain meaningful
+	 * GPIO output.
+	 */
+	static constexpr uint32_t output_mask()
+	{
+		return 0xFFFF'FFFF;
+	}
+
+	/**
+	 * The mask of bits of the `input` and `debouncedInput` registers that
+	 * contain meaningful GPIO input.
+	 */
+	static constexpr uint32_t input_mask()
+	{
+		return 0xFFFF'FFFF;
+	}
+
+	/**
+	 * Returns the bit corresponding to a given GPIO index. Will mask out
+	 * the bit if this is outside of the provided mask.
+	 */
+	static constexpr uint32_t output_enable_mask()
+	{
+		return 0xFFFF'FFFF;
+	}
+
+	/**
+	 * Returns the bit corresponding to a given GPIO index. Will mask out
+	 * the bit if this is outside of the provided mask.
+	 */
+	constexpr static uint32_t gpio_bit(uint32_t index, uint32_t mask)
+	{
+		return (1 << index) & mask;
+	}
+
+	/**
+	 * Set the output bit for a given GPIO pin index to a specified value.
+	 * This will only have an effect if the corresponding bit is first set
+	 * to `0` (i.e. output) in the `output_enable` register, and if the pin
+	 * is a valid output pin.
+	 */
+	void set_output(uint32_t index, bool value) volatile
+	{
+		const uint32_t Bit = gpio_bit(index, output_mask());
+		if (value)
+		{
+			output = output | Bit;
+		}
+		else
+		{
+			output = output & ~Bit;
+		}
+	}
+
+	/**
+	 * Set the output enable bit for a given GPIO pin index. If `enable` is
+	 * true, the GPIO pin is set to output. If `false`, it is instead set to
+	 * input mode.
+	 */
+	void set_output_enable(uint32_t index, bool enable) volatile
+	{
+		const uint32_t Bit = gpio_bit(index, output_enable_mask());
+		if (enable)
+		{
+			outputEnable = outputEnable | Bit;
+		}
+		else
+		{
+			outputEnable = outputEnable & ~Bit;
+		}
+	}
+
+	/**
+	 * Read the input value for a given GPIO pin index. For this to be
+	 * meaningful, the corresponding pin must be configured to be an input
+	 * first (set output enable to `false` for the given index). If given an
+	 * invalid GPIO pin (outside the input mask), then this value will
+	 * always be false.
+	 */
+	bool read_input(uint32_t index) volatile
+	{
+		return (input & gpio_bit(index, input_mask())) > 0;
+	}
+
+	/**
+	 * Read the debounced input value for a given GPIO pin index. FOr this
+	 * to be meaningful, the corresponding pin must be configured to be an
+	 * input first (set output enable to `false` for the given index). If
+	 * given an invalid GPIO pin (outside the input mask), then this value
+	 * will always be false.
+	 */
+	bool read_debounced_input(uint32_t index) volatile
+	{
+		return (debouncedInput & gpio_bit(index, input_mask())) > 0;
+	}
+};
+
+/**
+ * A driver for Sonata's Board GPIO (instance 0).
+ *
+ * Documentation source:
+ * https://lowrisc.org/sonata-system/doc/ip/gpio.html
```

Comment:
Perhaps examples in the documentation of the right `MMIO_CAPABILITY` invocation for these would be useful?  (Or even static factory methods, since they are backed by singletons in Sonata hardware?)

Reply:
This documentation page is meant to be purely a hardware reference as oppose to how-to/tutorial. The page does have a lot of room for improvement. We could also create a page with some minimal how-to guides that the drivers could point to, but I don't think this should live in the sonata-system book.

Reply:
I think static factory method / function would make sense.